proposals/extensions_v2.md

```diff
@@ -0,0 +1,349 @@
+3. Ensuring teams can move safely to modern extensions *especially* in environments where source binary compatibility is non-negotiable.
+
+However, there are parts of its core design that are not ideal which we would like to ensure we can expand on.  These expansions could be released with extensions if time and other resources permit.  Or they could came later and cleanly sit on top of the feature to improve the experience.
+
+These areas are:
+
+## Expansion 1: Syntactic clumsiness and repetition
+
+The initial extension form values source and binary compatibility as core requirements that must be present to ensure easy migration, allowing codebases to avoid both:
+1. bifurcation; where some codebases adopt modern extensions and some do not.
+2. internal inconsistency; where some codebases must keep around old extensions and new extensions, with confusion about the semantics of how each interacts with the other.
+
+Because classic extension methods have very few restrictions, modern extension methods need to be flexible enough to support all the scenarios supported there.
+
+However, many codebases do not need all the flexibility that classic extension methods afforded.  For example, classic extension methods allow disparate extension methods in a single static class to target multiple different types.  For use cases where that isn't required, we forsee a natural extension (pun intended) where one can modern translate extensions like so:
+
+```c#
+extension E
+{
+    // All extension members extend the same thing:
+
+    public void M() for SomeType { ... }
+    public int P for SomeType { get { ... } }
+    public static operator+(...) for SomeType { ... }
+    // etc
+}
+
+// Can be translated to:
+
+extension E for SomeType
+{
+    public void M() { ... }
+    public int P { get { ... } }
+    public static operator+(...) { ... }
+}
+```
+
+TODO: Do an ecosystem check on what percentage of existing extensions could use this simpler form.
+
+TODO: It's possible someone might have an extension where almost all extensions extend a single type, and a small handful do something slightly different (perhaps extending by `this ref`).  Would it be beneficial here to *still* allow the extension members to provide a `for-clause` to override that default for that specific member.  For example:
+
+```c#
+extension StringExtensions for string
+{
+    // Lots of normal extension methods on string ...
+
+    // Override here to extend `string?`
+    public bool MyIsNullOrEmpty() for [NotNullWhen(false)] string? str
+    {
+    }
+}
+```
+
+It seems like this would be nice to support with little drawback.
+
+## Expansion 2: Optional syntactic components
+
+As above, we want modern extensions to completely subsume classic extension methods.  As such, a modern extension  method must be able to support everything a classic extension method supported.  For example:
```

Comment:
Can we explicitly list out all of the patterns we want to support. The ones that come to mind for me are:

```csharp
for ref Span<T>
for ref readonly Span<T>
for in Span<T>
for scoped ref Span<T>
for scoped Span<T>
```

Think that is the exhaustive list.

Reply:
1.  Yes.  Happy to call that out.
2. Yes.  This proposal supports all those forms.

Reply:
i have called all thse forms out explicit above now.

Reply:
Is `scoped ref readonly` an interesting form here?